executor/src/witgen/jit/code_cleaner.rs

```diff
@@ -0,0 +1,153 @@
+use std::collections::HashSet;
+
+use powdr_number::FieldElement;
+
+use super::{
+    effect::Effect,
+    variable::{MachineCallVariable, Variable},
+};
+
+/// Returns the list of variables that are not needed to compute the requested
+/// variables.
+pub fn optional_vars<T: FieldElement>(
+    code: &[Effect<T, Variable>],
+    requested_variables: &[Variable],
+) -> HashSet<Variable> {
+    let mut required: HashSet<_> = requested_variables.iter().cloned().collect();
+    let mut optional: HashSet<_> = Default::default();
+    for effect in code.iter().rev() {
+        optional.extend(optional_vars_in_effect(effect, &mut required));
+    }
+    optional
+}
+
+/// Removes the given variables from the code and all variables that depend on them.
+pub fn remove_variables<T: FieldElement>(
+    code: Vec<Effect<T, Variable>>,
+    mut to_remove: HashSet<Variable>,
+) -> Vec<Effect<T, Variable>> {
+    code.into_iter()
+        .filter_map(|effect| remove_variables_from_effect(effect, &mut to_remove))
+        .collect()
+}
+
+/// Removes all calls to machines with the given IDs on the given row offsets.
+pub fn remove_machine_calls<T: FieldElement>(
+    code: Vec<Effect<T, Variable>>,
+    to_remove: &HashSet<(u64, i32)>,
+) -> Vec<Effect<T, Variable>> {
+    code.into_iter()
+        .filter_map(|effect| remove_machine_calls_from_effect(effect, to_remove))
+        .collect()
+}
+
+/// Returns the variables in the effect that are not needed
+/// to compute the requested variables and also updates the requested
+/// variables.
+/// This is intended to be used in reverse on a list of effects.
+fn optional_vars_in_effect<T: FieldElement>(
+    effect: &Effect<T, Variable>,
+    requested_variables: &mut HashSet<Variable>,
+) -> HashSet<Variable> {
+    let needed = match &effect {
+        Effect::Assignment(..) | Effect::ProverFunctionCall(..) => effect
+            .written_vars()
+            .any(|(v, _)| requested_variables.contains(v)),
+        Effect::Assertion(_) => {
+            // If none of the variables in the assertion are required, we can
+            // remove the assertion.
+            !effect
+                .referenced_variables()
+                .any(|v| requested_variables.contains(v))
```

Comment:
Does it hurt though? Can they become required later? I guess not, because we'd always do any assertion first before using the variable to derive more, so it would already be in the required set?